pkg/cmd/provisioning/azure/create_managed_identities.go

```diff
@@ -0,0 +1,467 @@
+	}
+	// Create a unique name for the role assignment
+	roleAssignmentName := uuid.String()
+	var rawResponse *http.Response
+	// Role assignment can fail due to a replication delay after creating the user-assigned managed identity
+	// Try up to 20 times with a 10 second delay each attempt
+	for i := 0; i < 20; i++ {
+		ctxWithResp := runtime.WithCaptureResponse(context.Background(), &rawResponse)
+		// armauthorization.RoleAssignmentsClientCreateResponse stomped because we don't need any information from
+		// the response however, ccoctl could potentially output the role assignment's ID similarly to other info
+		// which is log.Print'd
+		_, err = client.RoleAssignmentClient.Create(
+			ctxWithResp,
+			scope,
+			roleAssignmentName,
+			armauthorization.RoleAssignmentCreateParameters{
+				Properties: &armauthorization.RoleAssignmentProperties{
+					PrincipalID:      to.Ptr(managedIdentityPrincipalID),
+					RoleDefinitionID: targetRole.ID,
+				},
+			},
+			&armauthorization.RoleAssignmentsClientCreateOptions{},
+		)
+		if err != nil {
+			var respErr *azcore.ResponseError
+			if errors.As(err, &respErr) {
+				switch respErr.ErrorCode {
+				case "PrincipalNotFound":
+					// Replication delay. The identity we just created can't be found yet so we need to retry.
+					// TODO: Would it be better to display a message like this while we retry or would it be
+					//       acceptable to output a log line only once we've finished assigning the role?
+					if i == 19 {
+						log.Fatal("Timed out assigning role to user-assigned managed identity, this is most likely due to a replication delay following creation of the user-assigned managed identity, please retry")
+						break
+					} else {
+						log.Printf("Unable to assign role to user-assigned managed identity, retrying...")
+						time.Sleep(10 * time.Second)
+						continue
+					}
+				case "RoleAssignmentExists":
+					// Role assignment already present, continue
+					break
+				default:
+					return err
+				}
+			} else {
+				return err
+			}
+		} else {
+			break
+		}
+	}
+	log.Printf("Assigned %s role to user-assigned managed identity with principal ID %s with scope %s", roleName, managedIdentityPrincipalID, scope)
+	return nil
+}
+
+// createUserAssignedManagedIdentity creates a user-assigned managed identity with managedIdentityName.
+//
+// The user-assigned managed identity will be tagged with "Name": <managedIdentityName> as well as
+// any provided resourceTags.
+func createUserAssignedManagedIdentity(client *azureclients.AzureClientWrapper, managedIdentityName, resourceGroupName, region string, resourceTags map[string]string) (*armmsi.Identity, error) {
+	identityParameters := armmsi.Identity{
+		Location: to.Ptr(region),
+		Tags: map[string]*string{
+			nameTagKey: to.Ptr(managedIdentityName),
+		},
+	}
+	// Add provided tags to user-assigned managed identity parameters
+	for tagKey, tagValue := range resourceTags {
```

Comment:
Code review: should we check if the managed identities exist, validate that their tags are correct and conditionally update them to be correct (with merged tags) as we do with the resource group and storage account for symmetry? Probably.